### PR TITLE
Added testing framework

### LINKED ISSUE #28: [Team4] Make testing framework
Develop a simple testing framework. The testing framework should enable the following:
- [x] Provide general interface to define tests
- [x] Easily run all the tests
- [x] Include templating to switch between single and double precision
- [x] Properly handle errors being thrown while test cases are running
- [x] Handle displaying of results and returning proper value to indicate whether all tests passed or not

### PR DESCRIPTION
**Pull request by eytanadler**
Added a framework to define and run tests. This resolves #28.

### CODE REVIEW

File: test/readme.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+}
+```
+
+## Assert statements
+
+A number of assert statements are available to use in the tests:
+
+`ASSERT_EQUAL(first, second)`
+* If `first == second` evaluates to false, the test will fail.
+* `first` and `second` may be any type
```
Comment (akleb) [inline test/readme.md:41]: any type for which the `==` operator is defined? So use defined types do not work until an operator is defined.

File: test/readme.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+`ASSERT_NOT_EQUAL(first, second)`
+* If `first != second` evaluates to false, the test will fail.
+* `first` and `second` may be any type
```
Comment (akleb) [inline test/readme.md:45]: Same here

File: test/run_tests.sh [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+rm -rf "$f"
+touch "$f"
+echo "#include \"unit_test_framework.h\"" >> "$f"
+
+# Automatically include tests from any cpp file that starts with "test"
+for fTest in test*.cpp; do
+    printf "#include \"%s\"\n" "$fTest" >> "$f"
+done
+
+printf "\nTEST_MAIN()\n" >> "$f"
```
Comment (akleb) [inline test/run_tests.sh:12]: I thought you mentioned before that just including the different files with tests didn't work? Is this not the case

File: test/run_tests.sh [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+echo "=============================================================="
+echo "                    Single precision tests                    "
+echo "=============================================================="
+g++ -Wall -Werror -pedantic -O1 -std=c++11 -DFP_TYPE=float "$f" -o test_single.exe
+./test_single.exe
+printf "\n"
+echo "=============================================================="
+echo "                    Double precision tests                    "
+echo "=============================================================="
+g++ -Wall -Werror -pedantic -O1 -std=c++11 -DFP_TYPE=double "$f" -o test_double.exe
```
Comment (akleb) [inline test/run_tests.sh:24]: Why only O1?

File: test/run_tests.sh [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+./test_double.exe
+
+rm -rf test_single.exe
+rm -rf test_double.exe
```
Comment (akleb) [inline test/run_tests.sh:28]: Maybe we should leave executables around for debugging purposes? Possibly add debug flags to the test copmilation

File: test/run_tests.sh [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+rm -rf "$f"
```
Comment (akleb) [inline test/run_tests.sh:29]: If we have this at the beginning of the file, why add it here too? Seems like keeping it around for debugging would be useful.

File: test/run_tests.sh [hunk 7]
```diff
@@ -0,0 +1,29 @@
+# Build file with main to run all tests
+f="scratch_run_file.cpp"
+rm -rf "$f"
```
Comment (akleb) [inline test/run_tests.sh:3]: `-r` needed here?

**Review by akleb — CHANGES_REQUESTED** [review]

File: test/run_tests.sh (hunk 3 above)
Comment (eytanadler) [reply test/run_tests.sh:12]: I'm not sure exactly what changed but it's working now...

File: test/readme.md (hunk 1 above)
Comment (eytanadler) [reply test/readme.md:41]: Good point. This part was straight from the 280 docs...

File: test/run_tests.sh (hunk 4 above)
Comment (eytanadler) [reply test/run_tests.sh:24]: The EECS 280 profs like slow code

**Review by akleb — APPROVED** [review]
I am okay with this, but I am curious why no optimization flags with the debug flags? I typically do one or the other, but if both are possible, why not do both?

File: test/readme.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
+# Running tests
```
Comment (akleb) [inline test/readme.md:1]: This file needs to be updated for ctest

**Review by akleb — CHANGES_REQUESTED** [review]
Update README

**Pull request merged**